common/lib/plugins/bluegreen/blue_green_plugin.ts

```diff
@@ -0,0 +1,236 @@
+      }
+
+      const hostRole: BlueGreenRole = this.bgStatus.getRole(hostInfo);
+
+      if (!hostRole) {
+        // Connection to a host that isn't participating in BG switchover.
+        return this.regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      let client: ClientWrapper | null = null;
+      let routing: ConnectRouting | undefined = this.bgStatus.connectRouting.filter((routing: ConnectRouting) =>
+        routing.isMatch(hostInfo, hostRole)
+      )[0];
+
+      if (!routing) {
+        return this.regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      this.startTimeNano = getTimeInNanos();
+      while (routing && !client) {
+        client = await routing.apply(this, hostInfo, props, isInitialConnection, connectFunc, this.pluginService);
+        if (!client) {
+          this.bgStatus = this.pluginService.getStatus<BlueGreenStatus>(BlueGreenStatus, this.bgdId);
+          if (! this.bgStatus) {
+            this.endTimeNano = getTimeInNanos();
+            return this.regularOpenConnection(connectFunc, isInitialConnection);
+          }
+          routing = this.bgStatus.connectRouting.filter((routing: ConnectRouting) => routing.isMatch(hostInfo, hostRole))[0];
+        }
+      }
+
+      this.endTimeNano = getTimeInNanos();
+      if (!client) {
+        client = await connectFunc();
+      }
+
+      if (isInitialConnection) {
+        // Provider should be initialized after connection is open and a dialect is properly identified.
+        this.initProvider();
+      }
+
+      return client;
+    } finally {
+      if (this.startTimeNano > 0 && this.endTimeNano === BigInt(0)) {
+        this.endTimeNano = getTimeInNanos();
+      }
+    }
+  }
+
+  async execute<T>(methodName: string, methodFunc: () => Promise<T>, methodArgs: any[]): Promise<T> {
+    this.resetRoutingTimeNano();
+
+    try {
+      this.initProvider();
+
+      if (BlueGreenPlugin.CLOSED_METHOD_NAMES.has(methodName)) {
+        return await methodFunc();
+      }
+
+      this.bgStatus = this.pluginService.getStatus<BlueGreenStatus>(BlueGreenStatus, this.bgdId);
+
+      if (!this.bgStatus) {
+        this.endTimeNano = getTimeInNanos();
+        return await methodFunc();
+      }
+
+      const currentHostInfo: HostInfo = this.pluginService.getCurrentHostInfo();
+      const hostRole: BlueGreenRole = this.bgStatus.getRole(currentHostInfo);
+
+      if (!hostRole) {
+        // Connection to a host that isn't participating in BG switchover.
+        return await methodFunc();
+      }
+
+      let result: T | null = null;
+      let routing: ExecuteRouting | undefined = this.bgStatus.executeRouting.filter((routing: ExecuteRouting) =>
+        routing.isMatch(currentHostInfo, hostRole)
+      )[0];
+
+      if (!routing) {
+        return await methodFunc();
+      }
+
+      this.startTimeNano = getTimeInNanos();
+
+      while (routing && !result) {
```

Comment:
In JDBC `result` is an Optional so that we can distinguish between when the execute routing did not call the jdbc method (in which case the result will be Optional.EMPTY) versus when the method was called but it was void or returned null (in which case the result will be Optional.of(null)). Do we need to do the same here? We only want to enter this block if the method was not called, and !result evaluates to true if result is null or undefined